### PR TITLE
Implement basic transfer step in new woo installer

### PR DESCRIPTION
**Pull request by lsl**
#### Changes proposed in this Pull Request

* Implements a basic atomic transfer step in the new woo installer.

#### Testing instructions

You need a test site with a business plan on simple infra to begin this flow.

- http://calypso.localhost:3000/start/woocommerce-install/confirm?siteSlug=wooptestupgrade.wordpr\ess.com (change the site slug to your test site)
- Confirm the transfer ignoring the url change warning
- Transfer should eventually complete
- You should be redirected to wc-admin which automatically loads woo's onboarding "profile setup wizard"
- Your site should now be atomic with woo installed

To retest, use the revert tool in mc via blogrc, after the revert is triggered, it's good to refresh the page and click on the "Clean up Atomic Site Immediately" button to prevent any timing issues where you next test might get cleared by async job mid transfer.

Before

https://user-images.githubusercontent.com/811776/141722797-fb96270d-b2b5-435f-9d53-2078e5803c4b.mp4

After

https://user-images.githubusercontent.com/811776/141722826-4bf6a1d0-bb14-4b9c-b4f6-a91dce486183.mp4

### CODE REVIEW

File: client/signup/steps/woocommerce-install/transfer.tsx [hunk 1]
```diff
@@ -10,6 +22,76 @@ interface Props {
...
+		dispatch( initiateThemeTransfer( siteId, null, 'woocommerce' ) );
+	}, [ siteId, dispatch ] );
+
+	// Watch transfer status
+	useEffect( () => {
+		if ( fetchingTransferStatus ) {
+			return;
+		}
+
+		switch ( transferStatus ) {
```
Comment (lsl) [inline client/signup/steps/woocommerce-install/transfer.tsx]: todo: Don't like this much. The process is slow so we need something more than a spinner but I'm not sure this approach makes sense either.

Comment (lsl) [reply client/signup/steps/woocommerce-install/transfer.tsx]: Some prior art in here https://github.com/Automattic/wp-calypso/pull/56997/ uses some acceleration math to make things smoother.

Comment (lsl) [reply client/signup/steps/woocommerce-install/transfer.tsx]: Also, much better looking progress bar in here:

https://github.com/Automattic/wp-calypso/pull/57051/files#diff-d9bd8319fe61d2742e0a99a51a83f4f831414bae75fa6108c9e031f5da32a901R83-R111

(This PR included the new spinner graphic we should be using if required too)

File: client/signup/steps/woocommerce-install/transfer.tsx [hunk 2]
```diff
@@ -10,6 +22,76 @@ interface Props {
...
+			case transferStates.NONE:
+			case transferStates.PENDING:
+			case transferStates.INQUIRING:
+			case transferStates.PROVISIONED:
+			case transferStates.FAILURE:
```
Comment (retrofox) [inline client/signup/steps/woocommerce-install/transfer.tsx]: are `FAILURE`, `REVERTED`, `CONFLICTS` states that represent 20% of the whole process?

Comment (retrofox) [reply client/signup/steps/woocommerce-install/transfer.tsx]: oh, I see... it's handled below. 😶

File: client/signup/steps/woocommerce-install/transfer.tsx [hunk 3]
```diff
@@ -10,6 +22,76 @@ interface Props {
...
+			transferStatus === transferStates.ERROR ||
+			transferStatus === transferStates.FAILURE ||
+			transferStatus === transferStates.REQUEST_FAILURE ||
+			transferStatus === transferStates.CONFLICTS
+		) {
+			setProgress( 100 );
+			setError( { transferFailed, transferStatus } );
+		}
+	}, [ setProgress, goToStep, fetchingTransferStatus, transferStatus, transferFailed ] );
+
```
Comment (retrofox) [inline client/signup/steps/woocommerce-install/transfer.tsx]: Should we check whether the site ID is not defined? Pretty sure it is a really unlikely situation, but maybe it is worth it:

```es6
if ( ! siteId ) {
  return null;
}
```

File: client/signup/steps/woocommerce-install/transfer.tsx [hunk 4]
```diff
@@ -26,6 +108,9 @@ export default function Transfer( { goToStep }: Props ): ReactElement | null {
 			</div>
 			<div className="woocommerce-install__content">
 				<div>
+					{ error.transferFailed && 'error...' }
+					{ error.transferFailed && error.transferStatus }
```
Comment (retrofox) [inline client/signup/steps/woocommerce-install/transfer.tsx]: that's WIP, right?  😅

**Comment by allilevine** [comment]
@lsl @retrofox Other transfer flow components:
- The new Plugin Flow, which takes transfer status into account: #56997
- The ecommerce flow, which mostly just estimates the timing: #55073

It would be great if we could repurpose these or create shared components. 🤞

File: client/signup/steps/woocommerce-install/transfer.tsx (hunk 3 above)
Comment (lsl) [reply client/signup/steps/woocommerce-install/transfer.tsx]: Probably a good idea to be defensive. siteSlug is ensured by singup framework dependencies which sets siteId but I'm not too sure how solid that code is.

File: client/signup/steps/woocommerce-install/transfer.tsx (hunk 2 above)
Comment (lsl) [reply client/signup/steps/woocommerce-install/transfer.tsx]: > represent 20% of the whole process?

No this is something that needs improving, I just made it up for illustrative purposes. 

In practice you'll most likely only ever see states:
- Reverted
- Active
- Complete

You can potentially see more depending on timing of requests but its rare.

The actual request is extremely slow. Some discussion about the request speed here. p1634705094195400-slack-C7YPW6K40

File: client/signup/steps/woocommerce-install/transfer.tsx (hunk 4 above)
Comment (lsl) [reply client/signup/steps/woocommerce-install/transfer.tsx]: Yes, WIP, needs some work.

Actually managed to trigger this somehow, I like that the progress bar looks like it keeps attempting :smile: 

https://user-images.githubusercontent.com/811776/141056789-1362f4c2-dfa6-4c78-80e4-e0237fb02e2c.mp4

File: client/signup/steps/woocommerce-install/transfer/style.scss [hunk 5]
```diff
@@ -0,0 +1,59 @@
+@import '@automattic/onboarding/styles/mixins';
+
```
Comment (lsl) [inline client/signup/steps/woocommerce-install/transfer/style.scss:2]: Rushed this, could use tidying up, don't think we need all of it.

**Comment by lsl** [comment]
@allilevine I totally missed your comment this morning, looks like we had the same idea though! I suspect [my sources](https://github.com/Automattic/wp-calypso/pull/57811#discussion_r745328584) used your earlier work as their source :laughing: 

The result seems to be the same? Let me know if there are other improvements you think could be done here?

**Comment by allilevine** [comment]
> @allilevine I totally missed your comment this morning, looks like we had the same idea though! I suspect [my sources](https://github.com/Automattic/wp-calypso/pull/57811#discussion_r745328584) used your earlier work as their source 😆

Np I think I used code in the /new flow as my source 😆 

> The result seems to be the same? Let me know if there are other improvements you think could be done here?

I haven't been able to get the flow working today, I get a blank page with no console errors 🤔 But it looks like it's much more responsive to the transfer states, which is awesome.

**Comment by lsl** [comment]
@retrofox lets leave error handling & fixing plugin install for another PR. Keen to deploy this as is so I'm not blocking anyone with it.

Re error handling, I think we want to store the error in flow state and send them back to the confirm step with the error displayed but dismissable if that checks out?

edit: that was lucky, just ran into a timing issue where the transfer step completes and you're redirected to wc-admin, only the plugin install wasn't finished and I got hit with a 403. (plugin installs are async jobs)

I was expecting the `install` step or `complete` step to do a final fact checking before sending the user off to wc-admin so I don't think we need to worry about this but just wanted to raise it, added a ticket to the board so it doesn't get lost / forgotten.

File: client/signup/steps/woocommerce-install/transfer/index.tsx [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	goToStep: GoToStep;
+}
+
+export default function Transfer( { goToStep }: Props ): ReactElement | null {
+	const { __ } = useI18n();
+	const dispatch = useDispatch();
+
+	const [ progress, setProgress ] = useState( 0.1 );
+	const [ error, setError ] = useState( { transferFailed: false, transferStatus: null } );
+	const [ step, setStep ] = useState( __( 'Upgrading your site' ) );
```
Comment (a8ci18n) [inline client/signup/steps/woocommerce-install/transfer/index.tsx]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **18** times:
`translate( 'Upgrading your site…' )` *ES Score: 8*

File: client/signup/steps/woocommerce-install/confirm.tsx [hunk 7]
```diff
@@ -39,7 +38,7 @@ export default function Confirm( { goToStep }: Props ): ReactElement | null {
 		dispatch( requestEligibility( siteId ) );
 	}, [ siteId, dispatch ] );
 
-	const fetchingTransferStatus = !! useSelector( ( state: AppState ) =>
+	const fetchingTransferStatus = !! useSelector( ( state ) =>
```
Comment (retrofox) [inline client/signup/steps/woocommerce-install/confirm.tsx:41]: any particular reason why you removing this?

**Review by retrofox — APPROVED** [review]
LGTM.

https://user-images.githubusercontent.com/77539/141777474-1540db0a-6de0-4585-8d99-f76924c4efc2.mov

Let's keep tackling other tasks in follow-up implementations.

File: client/signup/steps/woocommerce-install/confirm.tsx (hunk 7 above)
Comment (lsl) [reply client/signup/steps/woocommerce-install/confirm.tsx:41]: I assume someone fixed the inference on something somewhere. Didn't dig too far but my linter stopped complaining so figured I'd clean them up.

**Comment by DustyReagan** [comment]
With a brand new Business plan, this Calypso PR sandboxed, and pointing to production wpcom, I noticed an "Installation in progress" warning when I first hit the install page. And when installation completed I got a "Sorry, you are not allowed to access this page." error during the redirect.

https://user-images.githubusercontent.com/140841/142031170-39ca9b42-7c73-4959-8de9-53bade836b42.mp4

**Comment by retrofox** [comment]
> With a brand new Business plan, this Calypso PR sandboxed, and pointing to production wpcom, I noticed an "Installation in progress" warning when I first hit the install page. And when installation completed I got a "Sorry, you are not allowed to access this page." error during the redirect.

Weird, but I see similar issues not only on this flow but in other ones, for instance, when installing a plugin on a simple site.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6977356

Hi @lsl, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include this string: `Last paint touchups`

Thank you in advance!

**Comment by lsl** [comment]
Thanks for merging!

The "sorry you're not allowed to access this page" and similar issues are down to timing, the plugin installs happen in an async job. This is something we need to poll for before redirecting.

I'm thinking either:
- Add a polling step to confirm plugin install before redirecting
- Add a polling step to the install step, and use the install step after the transfer step

I think it makes more sense to implement this all in the transfer step and leave the install step as a branch step for already atomic sites that just need to install or activate woo again.

cc @pablinos as we were discussing this yesterday some.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6977356).

**Pull request merged**